amp/class-amp-service-worker.php

```diff
@@ -39,6 +39,91 @@ public function init() {
 		add_action( 'wp_front_service_worker', array( $this, 'add_amp_runtime_caching' ) );
 		add_action( 'wp_front_service_worker', array( $this, 'add_image_runtime_caching' ) );
 		add_action( 'wp_front_service_worker', array( $this, 'add_live_list_offline_commenting' ) );
+
+		$theme_support = AMP_Theme_Support::get_theme_support_args();
+		if ( isset( $theme_support['app_shell'] ) ) {
+
+			// App shell is mutually exclusive with navigation preload.
+			add_filter( 'wp_service_worker_navigation_preload', '__return_false' );
+
+			// Opt to route all navigation requests through the app shell.
+			add_filter(
+				'wp_service_worker_navigation_route',
+				function () {
+					$template   = get_template();
+					$stylesheet = get_stylesheet();
+					$revision   = sprintf( '%s-v%s', $template, wp_get_theme( $template )->Version );
+					if ( $template !== $stylesheet ) {
+						$revision .= sprintf( ';%s-v%s', $stylesheet, wp_get_theme( $stylesheet )->Version );
+					}
+					$revision .= sprintf( ';user-%d', get_current_user_id() );
+
+					/*
+					 * Note that themes will need to vary the revision further by whatever is contained in the app shell.
+					 * In particular, the post_modified times of the nav menu items appearing in the header and footer should
+					 * be included in the revision.
+					 */
+					$revision .= ';' . md5(
```

Comment:
The calculation of the revision here is redundant with the login in the PWA plugin.